cfbs/core.py

```diff
@@ -0,0 +1,493 @@
+            return default
+        return self[key]
+
+    def _expand_index(self):
+        index = self._unexpanded
+        if type(index) in (dict, OrderedDict):
+            self._data = {"type": "index", "index": index}
+            return
+
+        assert type(index) is not Index
+        assert type(index) is str
+
+        self._data = get_or_read_json(index)
+
+        if not self._data:
+            sys.exit("Could not download or find module index")
+        if "index" not in self._data:
+            sys.exit("Empty or invalid module index")
+
+    @property
+    def data(self) -> dict:
+        if not self._data:
+            self._expand_index()
+        return self._data
+
+    def exists(self, module):
+        return os.path.exists(module) or (module in self)
+
+    def check_existence(self, modules):
+        for module in modules:
+            if not self.exists(module):
+                user_error("Module '%s' does not exist" % module)
+
+    def translate_aliases(self, modules):
+        translated = []
+        for module in modules:
+            if not module in self and os.path.exists(module):
+                translated.append(local_module_name(module))
+                continue
+            data = self[module]
+            if "alias" in data:
+                print("%s is an alias for %s" % (module, data["alias"]))
+                module = data["alias"]
+            translated.append(module)
+        return translated
+
+    def get_module_object(self, name, added_by=None):
+        module = OrderedDict({"name": name})
+        if name.startswith("./"):
+            object = _generate_local_module_object(name)
+        else:
+            object = self[name]
+        module.update(object)
+        if added_by:
+            module["added_by"] = added_by
+        module.move_to_end("steps")
+        return module
+
+
+def _construct_provided_module(name, data, url, url_commit):
+    module = OrderedDict()
+    module["name"] = name
+    if "description" not in data:
+        user_error(
+            "missing required key 'description' in module definition: %s" % pretty(data)
+        )
+    module["description"] = data["description"]
+    module["url"] = url
+    module["commit"] = url_commit
+    subdirectory = data.get("subdirectory")
+    if subdirectory:
+        module["subdirectory"] = subdirectory
+    dependencies = data.get("dependencies")
+    if dependencies:
+        module["dependencies"] = dependencies
+    if "steps" not in data:
+        user_error(
+            "missing required key 'steps' in module definition: %s" % pretty(data)
+        )
+    module["steps"] = data["steps"]
+    module["added_by"] = "cfbs add"
+    return module
+
+
+def _has_autorun_tag(filename):
+    assert os.path.isfile(filename)
+    content = read_file(filename)
+
+    return (
+        "meta:" in content
+        and "tags" in content
+        and "slist" in content
+        and "autorun" in content
+    )
```

Comment:
What happens if i have the following in a policy file?
```
# TODO: add autorun

bundle agent bogus
{
  meta:
      "tags" slist => { "deprecated" };
}
```

Reply:
This is probably outside the scope of this PR, but we should make a ticket for this if it is a problem.

Reply:
Yes, we don't want to implement policy parsing, so there are many ways to circumvent this check.